### DISCUSSION TITLE
mypy detects a dict as an object

### DISCUSSION THREAD

**Discussion by Sanchoyzer** [python/typing#1896] [Q&A] [answered]
I've noticed a bit of strange behaviour: mypy detects a dict as an object:

```python
def func_1() -> None:
    i: dict[str, str | None]
    for i in [{'a': 'qwerty'}, {'a': None}]:  # error: Incompatible types in assignment (expression has type "object", variable has type "dict[Any, Any]")  [assignment]
        print(i.get('a'))


def func_2() -> None:
    for i in [{'a': 'qwerty'}, {'a': None}]:
        print(i.get('a'))  # "object" has no attribute "get"  [attr-defined]


def func_3() -> None:
    my_list: list[dict[str, str | None]] = [{'a': 'qwerty'}, {'a': None}]  # it's ok
    for i in my_list:
        print(i.get('a'))
```

```
>>> reveal_type([{'a': 'qwerty'}, {'a': None}])
Runtime type is 'list'
[{'a': 'qwerty'}, {'a': None}]

>>> reveal_type([{'a': 'qwerty'}, {'a': None}][0])
Runtime type is 'dict'
{'a': 'qwerty'}

>>> reveal_type([{'a': 'qwerty'}, {'a': None}][1])
Runtime type is 'dict'
{'a': None}
```

Gist URL: https://gist.github.com/mypy-play/edab7242ef2538c33018bd2352baf38a
Playground URL: https://mypy-play.net/?mypy=latest&python=3.12&gist=edab7242ef2538c33018bd2352baf38a

And it's ok for [Pyright](https://pyright-play.net/?code=CYUwZgBGCuB2DGB9AjACgJQQLQD4IDkB7WEALgCgIqIBLUiYG%2BAFwG0BnZgJwBoJOuEAD4FiIALqVqYQoJq1YEVgG8A5AENV9VQEcA7iC7MAnqoC%2BfNZvpESZ8fQgBiCIa6z6ASQSEAtgAd1ZhoAIwAbEAgTfxB2BQh1dnYaAHNYXxBYZghUEAAPfy5Y5OIIAAtEqOMYiAAiQhCAKxAWWr4AN3UuGnVwyIq46MjaxhZWAEFYYz5J43FazCVE5LSMrMlqTYhCmizUGgA6FJBmVA1VdHRya9BIGAREACYMbDxbMikqGTl4lXPtfSGEzmSz-UR2ByfLY7PaHY6nc6XKguepNFrMWrlSqwQgJZjcULQZjDeGYpb4rhYW67EDASQ3cBQOBIADML1w4I%2Bm18xkQYRonHo-M4rFGbAEfAEwk54nEEAAvEorFoILoDEZTBYIMqbGJ7MjaMxVHFCABrKHfWjxHl8gXMChbagw05wk5nTSXIA)

**Comment by JelleZijlstra** [comment, marked as answer]
This is related to mypy's longstanding behavior of using the join operator instead of using unions: https://github.com/python/mypy/labels/topic-join-v-union . We'll likely change this eventually in at least some circumstances in mypy (python/mypy#12056), but it's a complex change to make.
